### PR TITLE
MD Tab Update

### PR DESCRIPTION
**Pull request by podraco**
### Description of Changes
this is a fix for some problems that surged from the last pull request, as well add some improvements and upgrades to the class.

Possible fix for #770 
Possible fix for #874 
Possible fix for #878 
Possible fix for #880 
Possible fix for #885

This is a big change for the class, it has been proven with every example code provided by the documentation and kitchen sink.

This updates the documentation for the new implemented methods.

Between the fixes there are:
- simplified the `remove_widget ` function from `MDTabs`
- `remove_widget ` now removes binds created on `add_widget` for new tabs.
- max width of every tabt's title label is set to `"360dp"`
- both kv and python instance creation reviewed. (small differences that affected its behavior on new widgets)
- added exception raises if something is not filled correctly or found.
- Added more specific Exceptions to allow developers know what exactly went wrong.
- Added basic `MDCarousel` Documentation.
- `MDTabsBase.text` is now deprecated, still has backward compatibility, but the use `tab_label_text` is now enforced and the use of this property will show a message in the kivy logger.
- Fixed reference bug in which, if the tab with the reference was pressed, the selected tab would be missing its state.
- Tab_bar now behaves on the content's width. 
If the width of the tab's inside is bigger than the scroll view, the `tab_bar` will be set to scrollable, the padding of the layout will be set to `["52dp", 0]` and allow scroll, if the width is lesser or equal to the scrollview width, the scroll then is locked, and the canvas padding is set to `[0,0]`,  `ScrollView.do_scroll_x` is set to `False` in this scenario.
- Some duplicated code was removed
- Some duplicated movements were removed, since some steps were already done by the children widget or by another event.    
- Fixed some small recursions.

Adds:
- tab_label_text namespace will store the raw tab's title label text. Replaces `MDTabsBase.text` property.
-  Tabs.switch_to is now more flexible, you can toggle between text, Tab_objects and if it's a string, search by its icon, title or raw text.
- Any text parsed to the Title property will be set with upper(). (this is to maintain every tab title clear)
- added title_is_capital, if the value is True, it will update the self. Text to be all capitals.

### CODE REVIEW

**Comment by podraco** [comment]
i have done multiple tests with different codes that where available on the issues reports and on the documentations and demos. i believe this changes works and fulfill the normal behavior of any kivyMD and comply with the Material IO standard.

If there's something that needs to be fixed, ill work on it as soon as possible as I can.

**Comment by podraco** [comment]
I'm waiting response from a collaborator, please don't approve merge yet.

**Review by runette — COMMENTED** [review]
As a general comment - I still have the problem that we are fundamentally changing the effect that `allow_stretch` has. This means that ALL previous apps that used `allow_stretch` MUST be changed and this does not even seem to be documented.

BEFORE (i.e. before about 2 weeks ago):

`allow_stretch=True` created the behavior described in material.io as "Fixed Tabs" - i.e. if possible it stretched the Tabs wider than the label text so that they filled the TabBar.

AFTER (i.e. now) :

`allow_stretch=True` means that the Tab will expand to the width of the text (i.e. pretty much what `allow_stretch = False` meant before.

I find it unfortunate and unnecessary to fundamentally reverse the meaning of an existing API element like this. It would be much better to deprecate the old behavior and create a new attribute to trigger the new behavior.

File: kivymd/uix/tab.py
Comment (runette) [inline kivymd/uix/tab.py:805]: This shold be compared to the actual and current font_style and not the imported one. Since TabLabel inherits from MDLabel it also inherits from ThemeableBehavior so this should be 

`if value in self.tab_label.theme_cls.font_styles`

File: kivymd/uix/tab.py
Comment (runette) [inline kivymd/uix/tab.py:523]: As a general comment - components should use the ThemableBehavior to get the current font_definitions and not import them - since they may have been changed by the application designer in the App.

File: kivymd/uix/tab.py
Comment (runette) [inline kivymd/uix/tab.py:851]: PLEASE - do not hard code the label title to Upper case.

This is a deal-breaker. There is NO specification in material.io that says that Tab labels should be all caps.

While you might think it is good practice and I might agree with you - this MUST be a matter for the app designer.

File: kivymd/uix/tab.py
Comment (runette) [inline kivymd/uix/tab.py:1203]: See the general comment about the change in the meaning of `allow_stetch`

In addition - although this mechanism works and is broadly in line with material.io - I do miss the functionality to ask kivyMD to expand the Tabs to fill the screen. It was useful functionality and was also in line with material.io

**Review by runette — CHANGES_REQUESTED** [review]
Starting to look good.

I do see some fairly fundamental bugs - as listed below

**Comment by runette** [comment]
General Comment :

When I run the new code in Kitchen Sink - there is a gap to the left of the first tab that I don't remember being there before. I don't have the time today to do a side-by-side analysis to confirm that. It does not look like a phantom tab because it cannot be selected.

Screenshot 2021-02-23 at 11 15 05

**Comment by runette** [comment]
General Comment :

When I run this new tab.py in my existing (released) application without other changes - I get this error message.

Throwing an error on legacy code that was working before could be a significant problem for the existing users of kivyMD.
```
builtins.ValueError: switch_tab:
name_tab not found in the tab list
search_by = 'text'
name_tab = 'Basic Details'
```

**Comment by HeaTTheatR** [comment]
@runette  Didn't know the labels were so different on `Mac OS` :)

Без имени

**Comment by runette** [comment]
> @runette Didn't know the labels were so different on `Mac OS` :)

I am using a custom font - mostly to test the most extreme case but also because my app uses the corporate brand font.

File: kivymd/uix/tab.py
Comment (podraco) [reply kivymd/uix/tab.py:805]: The documentation stated that was the proper use to register and read a new font. Let me test it in a different demo to see if that works, if not, a new issue must be opened

File: kivymd/uix/tab.py
Comment (podraco) [reply kivymd/uix/tab.py:851]: Well, the material io requieres that all labels must be clear, and. As far as I know, common practice to this is to use capital letters, since they have the most serif and use a bigger space to be drawn. But there's no problem, it can be removed

File: kivymd/uix/tab.py
Comment (podraco) [reply kivymd/uix/tab.py:1203]: It is a useful attribute, but, that shouldn't be it's name, since hat you're requiring, it's size_hint_x, from the grid layout that contains all tabs, implementing something to do the same in cpython to normal python is a really bad practice, but, i could forward the size hint of the tab bar to the MDTabsBase. So you could still use something like.
```py
tab_hint_x=1
tab_hint_x=None
```

File: kivymd/uix/tab.py
Comment (runette) [reply kivymd/uix/tab.py:805]: As is usual, the problem is actually in how you state the question!

In this code - you should not be trying to register or read a *new* font. You are trying to read the fonts and actually font_styles that *have been* registered with the app.

If you use the current code - you will *only* get the standard fonts provided out of the box. If the app developer has followed the procedure as documented to add a new font to their app - it will be ignored by your code. This is obviously a bug in this code. I do not see the need for a new issue.

**Comment by podraco** [comment]
That gap is required for scrollable tabs, it's in the specs and it's supposed to be `"52dp"`
If the width of every children in the container is lessed than the scroll view width, the do_scroll_x is False, and the padding =[0,0], this behavior updates every time the width is updated or a tab is added/removed/updated.

**Comment by podraco** [comment]
> General Comment :
> 
> When I run this new tab.py in my existing (released) application without other changes - I get this error message.
> 
> Throwing an error on legacy code that was working before could be a significant problem for the existing users of kivyMD.
> 
> ```
> builtins.ValueError: switch_tab:
> name_tab not found in the tab list
> search_by = 'text'
> name_tab = 'Basic Details'
> ```

That was my error, forgot to simplify the text than to use the property name as the default value

**Comment by podraco** [comment]
> General Comment :
> 
> When I run the new code in Kitchen Sink - there is a gap to the left of the first tab that I don't remember being there before. I don't have the time today to do a side-by-side analysis to confirm that. It does not look like a phantom tab because it cannot be selected.
> 
> Screenshot 2021-02-23 at 11 15 05

its because of this.

Captura de pantalla de 2021-02-23 15-14-06

Captura de pantalla de 2021-02-23 15-57-38

File: kivymd/uix/tab.py
Comment (podraco) [reply kivymd/uix/tab.py:523]: Well, as far as i know, when a dic is imported. The `__import__` method only creates a link to the pointer, and add the value to the global environment rather to contain it in the file/lib. Because of this, you could alter the value in any file and every change should be shared. Yet, I'm not certain if the event dispatcher would call any event to it.

Also, as far as I know, comparing a value to a list were always faster than to a dict.

Captura de pantalla de 2021-02-23 17-38-17
Captura de pantalla de 2021-02-23 17-37-22
Captura de pantalla de 2021-02-23 17-37-33
Captura de pantalla de 2021-02-23 17-39-16

File: kivymd/uix/tab.py
Comment (podraco) [reply kivymd/uix/tab.py:851]: yet, i just changed it to be optional.

File: kivymd/uix/tab.py
Comment (podraco) [reply kivymd/uix/tab.py:805]: The code only uses what's stored in memory.

File: kivymd/uix/tab.py
Comment (podraco) [reply kivymd/uix/tab.py:523]: The last commit ensures parity between the pointer and the theme_cls object.

**Comment by podraco** [comment]
So, these are the latest changes, fix some recursion calls by events, and adds what I believe was the last needed feature.
I'll wait until it can be reviewed.

I suggest that we merge this until everyone involved approves this PR.

File: kivymd/uix/tab.py
Comment (runette) [reply kivymd/uix/tab.py:805]: That is misdirection. It is all stored in memory. The question is whether you are using the correct values in the correct memory.

File: kivymd/uix/tab.py
Comment (runette) [reply kivymd/uix/tab.py:523]: To me, you are adding complexity on top of complexity and adding a lot of new code and possible error modes for the relatively small performance gain for getting the list of values from a list rather than dict items. 

The class `ThemeManager` HAS an existing attribute that holds the current font styles. Why not just always drive from the current value of that attribute rather than require another value to be kept up to date?

If I was technical lead on this project I would say this solution adds to the error vectors of the project and makes the project less maintainable - and indeed would probably not give this change "readability' approval. But I think this will work.

File: kivymd/theming.py
Comment (runette) [inline kivymd/theming.py:1067]: See comment below.

To me - this looks like a complex solution to fix a problem that does not really exist.

File: kivymd/uix/tab.py
Comment (runette) [reply kivymd/uix/tab.py:1203]: 1. It is not a matter of whether you think the name is a good name or whether I think it is a good name - IT IS THE EXISTING API. That is what existing application - some of them things that people make money out of - have included in their applications. Not because they are stupid or bad programmers but because that is what the previous specification WAS.

2. I did not suggest anything about implementation - so I am not sure why you seem to think that I was proposing to change something from cython to python? 

3. I am not sure that I understand the proposed solution - not enough info - but if it involves adding NEW complexity to an already complex API then probably not.

I think what I would really would have liked is:

- `allow_stretch` doing exactly what it did before (because :: legacy) and deprecated (because :: not really MD and badly named), and

- A "tab mode" attribute that has values of  "Fixed", "Scrollable" and "Fixed-length" for the various modes in material.io (or something similar to this).

File: kivymd/uix/tab.py
Comment (runette) [reply kivymd/uix/tab.py:851]: > As far as I know, common practice to this is to use capital letters, since they have the most serif and use a bigger space to be drawn

Thanks for making it optional.

Hard coding "good practice" is really not a good practice. And I do not necessarily agree with you. I use san-serif fonts (and Roboto is also san-serif) so that the question of the size of the serif has no relevance.

And if you look at from the point of view of psychology - all-caps words are harder for humans to read because we actually read by pattern recognition on the word and not by reading the letters.

**Comment by runette** [comment]
I have no further technical issues with this PR - it is working for me both in my app an in kitchen garden.

My only remaining concern is the breaking change in the API (i.e. the change in meaning and usaga of `allow_stretch`).

I will leave that to @HeaTTheatR to decide.

**Comment by runette** [comment]
BTW - just a procedural tip that you probably know already - if instead of 

"Possible fix for #770"
...

you had 

"fixes #770"
...

then merging the PR would automatically close the issue.

see https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

**Comment by syleishere** [comment]
Just tested this again, allow_stretch still won't work without tab_hint_x: True

**Comment by podraco** [comment]
Allow stretch is by defect `True`.
Allow stretch will allow the tab to stretch to the maximum size of the tab by the label content before wrapping it to 2 lines.
if `allow_stretch` is `False`, the tab won't change to the text texture, instead, it will use the fixed width property to set the tab's label width.

Meanwhile, `Tab_hint_x` will work outside the tab's texture size, and allow the tab to use a size hint based on the MDTabs widget width.

File: kivymd/theming.py
Comment (podraco) [reply kivymd/theming.py:1067]: Both namespaces should keep the same keys, values, otherwise, why should it exists as separated elements?

**Pull request merged**
